doc/cookbook/namedRoutes/NamedRoutes.lhs

```diff
@@ -0,0 +1,383 @@
+
+So, let's go deeper into our API tree.
+
+``` haskell
+data MovieAPI mode = MovieAPI
+  { get :: mode :- Get '[JSON] (Maybe Movie)
+  , update :: mode :- ReqBody '[JSON] Movie :> Put '[JSON] NoContent
+  , delete :: mode :- Delete '[JSON] NoContent
+  } deriving stock Generic
+```
+
+As you can see, we end up implementing the deepest routes of our API.
+
+Small detail: as our main API tree is also a record, we need the `NamedRoutes` helper.
+To improve readability, we suggest you create a type alias:
+
+``` haskell
+type MovieCatalogAPI = NamedRoutes API
+```
+
+That's it, we have our `MovieCatalogAPI` type!
+
+Let's make a server and a client out of it!
+
+## The Server
+
+As you know, we can’t talk about a server, without addressing the handlers.
+
+First, we take our handlers…
+
+```haskell
+versionHandler :: Handler Version
+versionHandler = pure "0.0.1"
+
+movieListHandler :: Maybe SortBy -> Handler [Movie]
+movieListHandler _ = pure moviesDB
+
+moviesDB :: [Movie]
+moviesDB =
+  [ Movie "1" "Se7en" 1995
+  , Movie "2" "Minority Report" 2002
+  , Movie "3" "The Godfather" 1972
+  ]
+
+getMovieHandler :: MovieId -> Handler (Maybe Movie)
+getMovieHandler requestMovieId = go moviesDB
+  where
+    go [] = pure Nothing
+    go (movie:ms) | movieId movie == requestMovieId = pure $ Just movie
+    go (m:ms) = go ms
+
+updateMovieHandler :: MovieId -> Movie -> Handler NoContent
+updateMovieHandler requestedMovieId newMovie =
+   -- update the movie list in the database...
+   pure NoContent
+
+deleteMovieHandler :: MovieId -> Handler NoContent
+deleteMovieHandler _ =
+   -- delete the movie from the database...
+   pure NoContent
+
+```
+
+And assemble them together with the record structure, which is the glue here.
+
+```haskell
+server ::  API AsServer
+server =
+  API
+    { version = versionHandler
+    , movies = moviesHandler
+    }
+
+moviesHandler :: MoviesAPI AsServer
+moviesHandler =
+  MoviesAPI
+    { list = movieListHandler
+    , movie = movieHandler
+    }
+
+movieHandler :: MovieId -> MovieAPI AsServer
+movieHandler movieId = MovieAPI
+    { get = getMovieHandler movieId
+    , update = updateMovieHandler movieId
+    , delete = deleteMovieHandler movieId
+    }
+```
+As you might have noticed, we build our handlers out of the same record types we used to define our API: `MoviesAPI` and `MovieAPI`. What kind of magic is this ?
+
+Remember the `mode` type parameter we saw earlier? Since we need to transform our API type into a _server_, we need to provide a server `mode`, which is `AsServerT Handler` here.
+
+You can alternatively use the AsServer (= AsServerT Handler) type alias. If you need to define handlers in some specific App monad from your codebase, the mode would simply be changed to AsServerT App.
+
+Finally, we can run the server and connect the API routes to the handlers as usual:
+
+``` haskell
```

Comment:
don't we want to use `genericServe` here? to stay in the "record world".

Reply:
There's no need as `NamedRoutes` has a `HasServer` instance

Reply:
I really don't see the point of introducing `genericServe` here — it makes the overall story less uniform.

We checked yesterday, and `genericServe` produces the exact same error message as `serve` if `Generic` instances are missing (a relatively unhelpful message mentioning `Generic`) so we are not gaining anything.

Reply:
Well that forces people to use `NamedRoutes` at the top, which is completely unnecessary if your entire API is made of records. The most direct way to serve a record of handlers is `genericServe`, hence my suggestion.

It'd be good to have some place that really explains those ":<|> separated" and "record" worlds and how we can go from one to the other, at all levels. But in the meantime, in the interest of giving the simplest code possible in the cookbook, without having to explain that named routes is an "adaptator" of sorts to plug records into old-style APIs, I'd recommend personally recommend `genericServe` (but do as you wish, this is just my opinion).

Reply:
I think:

```haskell
serveGeneric (Proxy @MyApi) myServer
```

isn't so much shorter than:

```haskell
serve (Proxy @(NamedRoutes MyAPI)) myServer
```

as to warrant the introduction of another function in the cookbook. It gives off the feeling that record APIs still somewhat live in a different world — which they now don't.

Reply:
That function is already introduced in the introductory cookbook: https://docs.servant.dev/en/stable/cookbook/generic/Generic.html#server

(and even the `ServerT` version)

And requires appending a few chars to `serve` vs another pair of parens and the `NamedRoutes` thing.

But I also suggested to @akhesaCaro at some point that another possibility could just be to pick one and show the equivalent with the other, in a commented-out (or not, could just be with a different name) line right below.